### DISCUSSION TITLE
useDeno Hook SSR sync vs async

### DISCUSSION THREAD

**Discussion by aadamsx** [alephjs/aleph.js#238] [General]
In the docs: https://alephjs.org/docs/advanced-features/use-deno-hook the Aleph way to fetch data is SSR or via a Client side async call out to an API.  But just reviewing the SSR examples:

```javascript
\\ sync
import React from "https://esm.sh/react";
import { useDeno } from "https://deno.land/x/aleph/mod.ts";

export default function Page() {
  const version = useDeno(() => {
    return Deno.version;
  });

  return <p>Powered by Deno v{version.deno}</p>;
}
```

vs

```javascript
\\ async
import React from "https://esm.sh/react";
import { useDeno, useRouter } from "https://deno.land/x/aleph/mod.ts";

export default function Post() {
  const { params } = useRouter();
  const post = useDeno(async () => {
    return await (await fetch(`https://.../post/${params.id}`)).json();
  });

  return <h1>{post.title}</h1>;
}
```

In the second example, even though it's async, isn't the process going to wait for the fetch from the database to return before pre-rendering and pushing the payload to the client?

**Comment by ije** [comment]
since it an asynchronous callback with deno runtime you can do any async task include database query

**Reply by ije** [reply]
but only in ssr mode

**Reply by aadamsx** [reply]
> since it an asynchronous callback with deno runtime you can do any async task include database query

Right, but what I'm saying is, it's async call to the db, but Aleph isn't going to pre-render and send the payload to the client before the callback returns anyhow right?  Aleph can go about making more calls, but in the end it's going to wait for all async calls to return from the db before it sends the payload back to the client.

**Reply by ije** [reply]
SSR just generate a html to the client and cache it until it is expired. and on the client the `useDeno` hook will be _tree-shared_ like:
```js
 const post = useDeno(null, 0, '#useDeno-hasufpa;q93w4u32');
```

**Comment by ije** [comment]
i perfer create a connection pool in `aleph.config.ts`, then use it in the hook

**Reply by aadamsx** [reply]
Do you have an example of this out there?

**Reply by ije** [reply]
will add a db example later
